File: next-app/src/lib/Clusterer.ts
/**
 * Clusterer.ts (V1.0 - Spatial Multi-Day Partitioning)
 * - Groups POIs into K days based on geographic proximity.
 * - Respects Hard Constraints: reservations MUST be in their assigned day.
 * - Soft Constraints: minimizes distance between spots in the same day.
 */

export interface ClusterablePlace {
  name: string;
  coords?: [number, number];
  reservation_date: string; // YYYY-MM-DD
}

export interface ClusteredDay {
  places: ClusterablePlace[];
  indices: number[]; // Original indices from the input array
}

/**
 * Partitions places into K days.
 * @param places The list of places to cluster
 * @param startDate The start date of the trip (YYYY-MM-DD)
 * @param numDays Total days in trip
 * @param baseCoords The starting coordinates (hotel/city center)
 */
export function clusterPlaces(
  places: ClusterablePlace[],
  startDate: string,
  numDays: number,
  baseCoords: [number, number] | null
): ClusteredDay[] {
  console.log(`[Clusterer V1.0] Partitioning ${places.length} places into ${numDays} days...`);

  // Initialize Clusters
  const clusters: ClusteredDay[] = Array.from({ length: numDays }, () => ({
    places: [],
    indices: []
  }));

  const tripStart = new Date(startDate + "T00:00:00");
  const unassignedIndices: number[] = [];

  // 1. Hard Reservation Assignment
  places.forEach((p, idx) => {
    if (p.reservation_date) {
      const resDate = new Date(p.reservation_date + "T00:00:00");
      const diffDays = Math.round((resDate.getTime() - tripStart.getTime()) / (1000 * 60 * 60 * 24));
      
      if (diffDays >= 0 && diffDays < numDays) {
        clusters[diffDays].places.push(p);
        clusters[diffDays].indices.push(idx);
        return;
      }
    }
    unassignedIndices.push(idx);
  });

  // 2. Greedy Proximity Assignment for "Free" Spots
  // We compute the centroid of each day (or use baseCoords if empty)
  unassignedIndices.forEach((idx) => {
    const p = places[idx];
    if (!p.coords) {
        // Fallback: assign to Day 1 if no coordinates (shouldn't happen with geocoding)
        clusters[0].places.push(p);
        clusters[0].indices.push(idx);
        return;
    }

    let minDistance = Infinity;
    let bestDay = 0;

    clusters.forEach((day, dayIdx) => {
      // Centroid of the day
      const centroid = calculateCentroid(day.places, baseCoords);
      const dist = calculateDistance(p.coords!, centroid);
      
      // Load Balancing Adjustment: 
      // Bias slightly away from already full days to keep schedule balanced
      const balancePenalty = day.places.length * 0.1; 
      const adjustedDist = dist + balancePenalty;

      if (adjustedDist < minDistance) {
        minDistance = adjustedDist;
        bestDay = dayIdx;
      }
    });

    clusters[bestDay].places.push(p);
    clusters[bestDay].indices.push(idx);
  });

  // Final confirmation
  clusters.forEach((day, i) => {
    console.log(`[Clusterer V1.0] Day ${i+1}: ${day.places.length} spots assigned.`);
  });

  return clusters;
}

function calculateCentroid(places: ClusterablePlace[], fallback: [number, number] | null): [number, number] {
  const activePlaces = places.filter(p => p.coords);
  if (activePlaces.length === 0) return fallback || [0, 0];

  let sumLat = 0, sumLon = 0;
  activePlaces.forEach(p => {
    sumLat += p.coords![0];
    sumLon += p.coords![1];
  });
  return [sumLat / activePlaces.length, sumLon / activePlaces.length];
}

function calculateDistance(c1: [number, number], c2: [number, number]): number {
  // Simple Euclidean distance for clustering (good enough for city-scale)
  const dy = c1[0] - c2[0];
  const dx = c1[1] - c2[1];
  return Math.sqrt(dx * dx + dy * dy);
}

File: api/engine/clusterer.py
import math
from typing import List, Dict, Tuple, Optional
from datetime import datetime

class ClusterablePlace(dict):
    name: str
    coords: Optional[Tuple[float, float]]
    reservation_date: Optional[str] # YYYY-MM-DD
    is_reservation: bool

class ClusteredDay(dict):
    places: List[ClusterablePlace]
    indices: List[int]

def calculate_centroid(places: List[ClusterablePlace], fallback: Tuple[float, float]) -> Tuple[float, float]:
    active_places = [p for p in places if p.get("coords")]
    if not active_places:
        return fallback

    sum_lat = sum(p["coords"][0] for p in active_places)
    sum_lon = sum(p["coords"][1] for p in active_places)
    return (sum_lat / len(active_places), sum_lon / len(active_places))

def calculate_distance(c1: Tuple[float, float], c2: Tuple[float, float]) -> float:
    # Simple Euclidean distance for clustering (city-scale)
    dy = c1[0] - c2[0]
    dx = c1[1] - c2[1]
    return math.sqrt(dx * dx + dy * dy)

def cluster_places(
    places: List[ClusterablePlace],
    start_date: str,
    num_days: int,
    base_coords: Optional[Tuple[float, float]] = None
) -> List[ClusteredDay]:
    """
    Python Implementation of greedy centroid-based spatial partitioning.
    :param places: List of places with coords and reservation_date
    :param start_date: Trip start date (YYYY-MM-DD)
    :param num_days: Total days in trip
    :param base_coords: Anchor point (e.g., hotel/city center)
    """
    clusters = [{"places": [], "indices": []} for _ in range(num_days)]

    active_places_with_coords = [p for p in places if p.get("coords")]
    
    # Calculate global centroid as fallback
    global_centroid = calculate_centroid(active_places_with_coords, base_coords or (0, 0))

    trip_start = datetime.fromisoformat(start_date)
    unassigned_indices: List[int] = []

    # 1. Hard Reservation Assignment
    for idx, p in enumerate(places):
        if p.get("reservation_date") and p.get("is_reservation"):
            try:
                res_date = datetime.fromisoformat(p["reservation_date"])
                diff_days = (res_date - trip_start).days
                if 0 <= diff_days < num_days:
                    clusters[diff_days]["places"].append(p)
                    clusters[diff_days]["indices"].append(idx)
                    continue
            except ValueError:
                pass
        unassigned_indices.append(idx)

    # 2. Initial Seeding for Empty Days
    # Give empty days one unassigned spot to anchor its centroid
    for d in range(num_days):
        if len(clusters[d]["places"]) == 0 and unassigned_indices:
            seed_idx = unassigned_indices.pop(0)
            clusters[d]["places"].append(places[seed_idx])
            clusters[d]["indices"].append(seed_idx)

    # 3. Greedy Proximity Assignment for remaining spots
    for idx in unassigned_indices:
        p = places[idx]
        if not p.get("coords"):
            # Fallback for places without coords
            clusters[0]["places"].append(p)
            clusters[0]["indices"].append(idx)
            continue

        min_distance = float('inf')
        best_day = 0

        for day_idx, day in enumerate(clusters):
            # Centroid of the day
            centroid = calculate_centroid(day["places"], base_coords or global_centroid)
            dist = calculate_distance(p["coords"], centroid)
            
            # Load Balancing Adjustment (Imbalance Penalty)
            imbalance_penalty = len(day["places"]) * 0.005 # ~500m penalty per item
            adjusted_dist = dist + imbalance_penalty

            if adjusted_dist < min_distance:
                min_distance = adjusted_dist
                best_day = day_idx

        clusters[best_day]["places"].append(p)
        clusters[best_day]["indices"].append(idx)

    return clusters
